Sprint 3/step 1

Neverhood93 opened the pull request:
Сверстаны дополнительные страницы связанные с аутентификацией и авторизацией.
Настроена маршрутизация, добавлен компонент для защиты маршрутов.
Добавлена страница профиля, настроена возможность изменения данных пользователя.
Создан новый сервис auth, для общения с api сервера.

Neverhood93 committed:
Создана папка pages.
Создан компонент страницы Home, в этот компонент вынесена основная логика компонента App.
Добавлен Preloader.

Neverhood93 committed:
В проект добавлена библиотека react-router-dom.

Neverhood93 committed:
Добавлен роутер и первый маршрут для Home.

Neverhood93 committed:
Добавлены заглушки для страниц и маршруты для них.

Neverhood93 committed:
Возвращена логика получения списка ингредиентов на App.
Добавлен маршрут для модального окна ингредиента.
Изменена логика работы модального окна, чтобы отрабатывала маршрутизация.
Добавлен маршрут и страница NotFound.

Neverhood93 committed:
Рефакторинг связанный с модальными окнами.

Neverhood93 committed:
Сверстана страница LoginPage без логики.

Neverhood93 committed:
Сверстана RegisterPage без логики.

Neverhood93 committed:
Сверстана ResetPasswordPage без логики.

Neverhood93 committed:
Сверстана ForgotPasswordPage без логики.

Neverhood93 committed:
Сверстана страница профиля без логики.
Добавлена заглушка для списка заказов.

Neverhood93 committed:
Добавлены конечные точки для login/register/logout и проведено тестовое подключение к ним.

Neverhood93 committed:
Доработано auth-api.

Neverhood93 committed:
Реализован черновой вариант функционала страницы регистрации.

Neverhood93 committed:
Реализован черновой вариант функционала страницы входа.

Neverhood93 committed:
Логика обработки события изменения полей вынесена в кастомный хук useForm.

Neverhood93 committed:
Добавлен вывод модального окна для отображения ошибки от сервера.

Neverhood93 committed:
Рефакторинг authSlice.

Neverhood93 committed:
Добавлен компонент ProtectedRoute для защиты маршрутов.
Изменен header для работы с навигацией.
Маршруты помещены в UnAuthRoute и AuthRoute соответственно.
Добавлен checkUserAuth для получения признака авторизации пользователя.

Neverhood93 committed:
Доработан функционал страниц ForgotPasswordPage, ResetPasswordPage.
Правки по auth-api.

Neverhood93 committed:
Исправлена ошибка, чтобы пароль не обновлялся, если значение поля не заполняли.

Neverhood93 committed:
Добавлена проверка на то, чтобы пользователь был авторизирован, когда нажимает на оформление заказа.

gennady-bars commented on src/utils/auth-api.tsx:
```diff
@@ -0,0 +1,106 @@
+    headers: baseApiConfig.headers,
+    body: JSON.stringify(requestData),
+  });
+  return getResponse(res, (data) => data as LoginResponse);
```
**Можно лучше**

Можно сделать **универсальную** функцию запроса с **проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(getResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request` и убрать дублирование проверки на `ok`. Все остальное будет без изменений. Код станет чище

И даже можно поместить внутрь `baseUrl`, чтобы не дублировать его в каждом запросе. Тогда нужно будет передавать просто `endpoint` в вызов

gennady-bars commented on src/services/auth/action.tsx:
```diff
@@ -0,0 +1,103 @@
+} from "../../utils/auth-api";
+import { AppDispatch } from "../store";
+
+export const register = createAsyncThunk<LoginResponse, RegisterRequest>(
+  "auth/register",
+  async (requestData) => {
+    return await registerEndpoint(requestData);
+  },
+);
+
+export const login = createAsyncThunk<LoginResponse, LoginRequest>(
+  "auth/login",
+  async (requestData) => {
+    return await loginEndpoint(requestData);
+  },
+);
+
+export const logout = createAsyncThunk<CommonResponse, string>(
+  "auth/logout",
+  async (refreshToken) => {
+    return await logoutEndpoint(refreshToken);
+  },
```
**Можно лучше**

* Если внутри **безымянной** (стрелочной) функции вызывается одна функция с **точно такими же** аргументами, то эта безымянная функция не нужна. Это лишняя обертка вокруг самой функции, которую можно использовать просто сразу в коде


```js

export const register = createAsyncThunk<LoginResponse, RegisterRequest>(
  "auth/register",
  registerEndpoint
);

export const login = createAsyncThunk<LoginResponse, LoginRequest>(
  "auth/login",
  loginEndpoint
);

export const logout = createAsyncThunk<CommonResponse, string>(
  "auth/logout",
  logoutEndpoint
);

export const forgotPassword = createAsyncThunk<
  CommonResponse,
  ForgotPasswordRequest
>("auth/forgotPassword", forgotPasswordEndpoint);

export const resetPassword = createAsyncThunk<
  CommonResponse,
  ResetPasswordRequest
>("auth/resetPassword", resetPasswordEndpoint);
```

gennady-bars commented on src/pages/login/loginPage.tsx:
```diff
@@ -0,0 +1,97 @@
+import { closeAuthModal } from "../../services/auth/reducer";
+
+function LoginPage() {
+  const { formState, handleFieldChange } = useForm<LoginRequest>({
```
Отлично, что используете хук `useForm`

gennady-bars commented on src/pages/login/loginPage.tsx:
```diff
@@ -0,0 +1,97 @@
+
+  return (
+    <div className={styles.container}>
+      <form className={styles.form} onSubmit={handleSubmit}>
```
Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. Обработка события `submit` еще срабатывает при нажатии на `Enter`, так как это встроенная логика сабмитов в браузере.

gennady-bars commented on src/pages/ingredient-details/ingredient-details.tsx:
```diff
@@ -0,0 +1,15 @@
+
+function IngredientDetailsPage() {
+  return (
+    <>
```
`<>` пустые теги нужны только тогда, когда нет **общего** тега (компонента), который оборачивает всю верстку

gennady-bars commented on src/pages/login/loginPage.tsx:
```diff
@@ -0,0 +1,97 @@
+    dispatch(closeAuthModal());
+  };
+
+  const handleSubmit = (e: FormEvent) => {
```
Хорошо типизировали функции

gennady-bars commented on src/components/protected-route/protected-route.tsx:
```diff
@@ -0,0 +1,40 @@
+    return <Preloader />;
+  }
+
+  if (onlyUnAuth && user) {
+    const { from } = (location.state as { from: { pathname: string } }) || {
+      from: { pathname: "/" },
+    };
+    return <Navigate to={from} />;
+  }
+
+  if (!onlyUnAuth && !user) {
+    return <Navigate to="/login" state={{ from: location }} />;
```
Отлично сделан `ProtectedRoute`

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -21,25 +31,74 @@ function App() {
 
+  const location = useLocation();
+  const navigate = useNavigate();
+  const background = location.state && location.state.background;
```
Отлично сделана логика проверки, был ли пользователь на сайте, и показывать ли страницу  с деталями, а не модально окно

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично, что используете хук `useForm`
* Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. 
* Хорошо типизировали функции
* Отлично сделан `ProtectedRoute`
* Отлично сделана логика проверки, был ли пользователь на сайте, и показывать ли страницу  с деталями, а не модально окно


но есть некоторые недочеты:



* `<>` пустые теги нужны только тогда, когда нет **общего** тега (компонента), который оборачивает всю верстку
* Пользователь не должен **напрямую** попасть на маршрут `/reset-password`. На эту страницу **никакой** пользователь не должен попадать напрямую. Только после нажатия на `восстановить пароль` нужно делать переход на эту страницу. Для этого нужно сделать проверку, был ли пользователь на странице `forgot-password` перед тем, как попасть на `reset-password` 
* `app.test.tsx` нужно удалить, так как он не относится к проекту  



**Можно лучше**

* Если внутри **безымянной** (стрелочной) функции вызывается одна функция с **точно такими же** аргументами, то эта безымянная функция не нужна. Это лишняя обертка вокруг самой функции, которую можно использовать просто сразу в коде
* при клике на **логотип** обычно идет переход на **главную** страницу
* Ошибки яндекс-библиотеки можно заглушить, чтобы типизация не ругалась в консоли

Neverhood93 committed:
Исправлено:
Пользователь не должен напрямую попасть на маршрут /reset-password. На эту страницу никакой пользователь не должен попадать напрямую. Только после нажатия на восстановить пароль нужно делать переход на эту страницу.
Удален файл app.test.tsx
<> пустые теги нужны только тогда, когда нет общего тега (компонента), который оборачивает всю верстку
Cделана универсальная функция запроса с проверкой ответа, чтобы не дублировать эту проверку в каждом запросе.
Если внутри безымянной (стрелочной) функции вызывается одна функция с точно такими же аргументами, то эта безымянная функция не нужна. Это лишняя обертка вокруг самой функции, которую можно использовать просто сразу в коде
Клик по лого, теперь возвращает на главный экран.
Убраны ненужные пропсы onPointerEnterCapture и onPointerLeaveCapture и добавлена заглушка {/* @ts-ignore */}, чтобы не захламлять консоль в браузере. А также ошибки с PasswordInput.

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.




* В булках не нужны `DragIcon` по макету https://disk.yandex.ru/i/MO-kXo5CcxFbLg

Neverhood93 committed:
Исправлено:
Убран компонент DragIcon у булок в конструкторе.
